### ISSUE TITLE
This panorama is too big for this device

### ISSUE THREAD

**Issue by edent** [mpetroff/pannellum#444]
On FireFox for Android, I'm able to view a large panorama.  This fails on Chrome & Samsung browser for Android, and iOS.

The error is:

> This panorama is too big for your device! It's 5660px wide, but your device only supports images up to 4096px wide. Try another device. (If you're the author, try scaling down the image.)

The 5660px images are natively generated by the LG 360.  Scaling the images works.

Am I right to assume from https://github.com/mpetroff/pannellum/blob/master/src/js/pannellum.js#L1535 that this is a WebGL limitation rather than a limitation with Pannellum?

(Brilliant in all other respects - so easy to use!)

**Comment by mpetroff** [comment]
Yes, it's a WebGL limitation (that depends on the browser/device/graphics driver combination).

**Comment by PaulKsi** [comment]
[link pannellum](https://codepen.io/AptusTim/pen/xMoqdW)
panolens js with the same panorama probably not webgl limitation? 
[link panolens](https://codepen.io/paulksi/pen/oNGQmww)

**Comment by jsohndata** [comment]
i fixed the issue by bringing down the image to 4,090px. saved it as WebP and brought the image from 2.5MB to about 600kb.

https://jsohndata.com/

**Comment by strarsis** [comment]
Using multires should avoid this issue, and also improve network bandwidth usage.

**Comment by Manish-Sataiva** [comment]
> i fixed the issue by bringing down the image to 4,090px. saved it as WebP and brought the image from 2.5MB to about 600kb.
> 
> https://jsohndata.com/

Image quality will be affected here right? is there any other solution to solve this issue?

**Comment by StigNygaard** [comment]
> Image quality will be affected here right? is there any other solution to solve this issue?

You could check the browser for supported size, and have two or more sizes to choose from depending on size supported.

I'm pretty new to Pannellum, being forced to look for a new way to display my equirectangular images after RoundMe closed without warning.

I prefer keeping it simple and just point a "viewer" to my equirectangular images. However I also discovered that not all browsers/systems can display my typically 16k wide equirectangular images. But I found out how to detect the maximal supported size of webGL. It does however makes thing less simple, because it requires me to make multiple sizes available of most of my equirectangular images. So, some might say it is still worth considering another solution than using equirectangular images directly...

It is some time ago I made it, and currently I don't remember every details of what I made (I think there's a bit of functionality in the backend, but the detection-part I believe is all in the front-end javascript code). But if you open [this page](https://www.rockland.dk/360/pannellum/?title=Reffen&image=2020Reffen/Reffen-20200807-2023&image8k=2020Reffen/Reffen-20200807-2023.8k&image16k=2020Reffen/Reffen-20200807-2023.16k&allowAvif=1) and look in the browser console, it will show something like:
```
Looking for features: avif,webgl
Detecting avif support... 
Detecting WebGL support and MaxSize... 
Found support: {"webGL":true,"webGLMaxSize":16384,"avif":true}
Setting up Pannellum with image: /360/img/2020Reffen/Reffen-20200807-2023.16k.avif
```
All the relevant client-side code creating this, is visible directly on page, if you do a view-source of the page.

I can try refresh my memory for more details, if interested?

**Comment by lborgman** [comment]
I am trying to find out how to check the image before loading it by pannellum (so I can fix the image). I followed the link in the comment by @edent , but it looks like it is no longer the line to look at:

https://github.com/mpetroff/pannellum/issues/444#issue-247842688

I would be glad to know the line number (or even more for how to check the image).

**Comment by mpetroff** [comment]
> I would be glad to know the line number (or even more for how to check the image).

That referred to:
https://github.com/mpetroff/pannellum/blob/8cb5902e1eca3d78994c46a8151f2646d49b36e7/src/js/pannellum.js#L1535

Which is now line 1801:
https://github.com/mpetroff/pannellum/blob/2d8ab26c360fdf1a28c00b791223b522c33dff9c/src/js/pannellum.js#L1801

> But where do I find the limits for WebGL?

See this comment: https://github.com/mpetroff/pannellum/issues/472#issuecomment-330059690

However, since v2.5.0, equirectangular images will be split into two textures if necessary, so images up to 8192px wide are supported by all devices. As this already corresponds to a 33.5 megapixel image, I'd recommend using Pannellum's [multiresolution feature](https://pannellum.org/documentation/examples/multiresolution/) for images that are larger than this, anyway.
